lib/services/elasticsearch.js

```diff
@@ -2029,42 +2007,35 @@ class ElasticSearch extends Service {
 
   /**
    * Scroll index in elasticsearch and return all document that match the filter
+   * /!\ throws a write_limit_exceed error: this method is intended to be used
+   * by deleteByQuery
    *
    * @param {Object} esRequest - Search request body
    *
    * @returns {Promise.<Array>} resolve to an array of documents
    */
-  _getAllDocumentsFromQuery ({ index, body, scroll, from, size }) {
-    const
-      client = this._client,
-      documents = [];
+  async _getAllDocumentsFromQuery (esRequest) {
+    let { body: { hits, _scroll_id } } = await this._client.search(esRequest);
 
-    return new Bluebird((resolve, reject) => {
-      client.search(
-        { index, body, scroll, from, size },
-        function getMoreUntilDone(error, { body: { hits, _scroll_id } }) {
-          if (error) {
-            return reject(error);
-          }
+    if (hits.total.value > this._kuzzle.config.limits.documentsWriteCount) {
+      errorsManager.throw('write_limit_exceeded');
+    }
 
-          for (const hit of hits.hits) {
-            documents.push({
-              _id: hit._id,
-              _source: hit._source
-            });
-          }
+    const documents = hits.hits.map(h => ({ _id: h._id, _source: h._source }));
 
-          if (hits.total.value !== documents.length) {
-            client.scroll({
-              scrollId: _scroll_id,
-              scroll
-            }, getMoreUntilDone);
-          }
-          else {
-            resolve(documents);
-          }
-        });
-    });
+    while (hits.total.value !== documents.length) {
+      ({ body: { hits, _scroll_id } } = await this._client.scroll({ //NOSONAR
+        scrollId: _scroll_id,
+        scroll: esRequest.scroll
+      }));
+
+      hits.hits.forEach(h => documents.push({
```

Comment:
`Array.map` with `Array.concat` is faster than `Array.forEach` and `Array.push`
```
   foreach & push x 105,624 ops/sec ±2.96% (86 runs sampled)
   for (const of) & push x 112,869 ops/sec ±1.38% (87 runs sampled)
   for (let i) & push x 110,137 ops/sec ±1.68% (86 runs sampled)
   map & concat x 137,630 ops/sec ±1.49% (81 runs sampled)
 Fastest is map & concat
```
https://gist.github.com/Aschen/bf4179b78f2be7fceae446419a442169